pandas/_libs/interval.pyi

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+import sys
+from typing import (
+    Any,
+    Generic,
+    Protocol,
+    TypeVar,
+    Union,
+    overload,
+)
+
+import numpy as np
+
+from pandas._typing import (
+    Timedelta,
+    Timestamp,
+)
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+VALID_CLOSED: frozenset
+
+OrderableScalarT = TypeVar("OrderableScalarT", int, float)
+OrderableTimesT = TypeVar("OrderableTimesT", Timestamp, Timedelta)
+OrderableT = TypeVar("OrderableT", int, float, Timestamp, Timedelta)
+
+class IntervalMixinProtocol(Protocol): ...
+
+class _LengthDescriptor:
+    @overload
+    def __get__(self, instance: Interval[float], owner: Any) -> float: ...
+    @overload
+    def __get__(self, instance: Interval[int], owner: Any) -> int: ...
+    @overload
+    def __get__(self, instance: Interval[OrderableTimesT], owner: Any) -> Timedelta: ...
+
+class IntervalMixin:
+    @property
+    def closed_left(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def closed_right(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def open_left(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def open_right(self: IntervalMixinProtocol) -> bool: ...
+    @property
+    def mid(self: IntervalMixinProtocol) -> float: ...
+    @property
+    def is_empty(self: IntervalMixinProtocol) -> bool: ...
```

Comment:
Can you explain why you need to add `IntervalMixinProtocol` everywhere? I'm actually surprised this is actually allowed (any class implements this empty protocol, when using TypeVars pyright is picky about that, see https://github.com/microsoft/pyright/issues/3088#issuecomment-1046335018)

Reply:
YOu're right.  Moved it to be in the class definition of `IntervalMixin`